File: python/013_reverseParentheses.py
def reverseParentheses(s):    
    '''
    You have a string s that consists of English letters, punctuation marks, whitespace characters, and brackets. 
    It is guaranteed that the parentheses in s form a regular bracket sequence.
    
    Your task is to reverse the strings contained in each pair of matching parentheses, starting from the innermost pair.    
    The results string should not contain any parentheses.
    '''
    l = list(s)
    stack = []
    for i in range(len(l)):
        if l[i] == '(':
            # opens a parenthesis, we must reverse this at some point
            stack.append(i)
        elif l[i] == ')':
            # close a parenthesis, reverse this from the last opening parenthesis found
            start = stack.pop()
            l[start + 1:i] = l[start + 1:i][::-1]
    return ''.join([i for i in l if (i != ')' and i != '(' )])

print(reverseParentheses("a(bc)de"))

File: arcade/go/arcade-intro/03_Smooth Sailing/013_reverseParentheses.go
import "strings"

func reverseInParentheses(s string) string {
	var stack []int
	var start, end int
	l := strings.Split(s, "")

	for i, item := range l {
		if item == "(" {
			// we found the start of a sequence, push it to the stack
			stack = append(stack, i)
		} else if item == ")" {
			// we found the end of a sequence, we need to pop the start of this sequence and reverse this
			end = i
			start, stack = stack[len(stack)-1], stack[:len(stack)-1]
			// Add 1 to start, and substract 1 to end, because we don't need to reverse the parenthesis
			reversePartOfArray(start+1, end-1, l)
		}
	}

	var result []string
	for _, item := range l {
		if item != "(" && item != ")" {
			result = append(result, item)
		}
	}

	return strings.Join(result, "")
}

func reversePartOfArray(start int, end int, l []string) {
	for i, j := start, end; i < j; i, j = i+1, j-1 {
		l[i], l[j] = l[j], l[i]
	}
}